### ISSUE TITLE
refactor the current tests

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#456]
Hi all,

I'm all for the integration tests they're useful and powerful but can we consider dropping the parallel test at least to a nightly build slot yet? (this at least saves us about ~5min on the tests).

Waiting 20min just to get in 4 lines of code seems a bit expensive especially when we have multiple very small PR waiting.

Does anyone have any opinion on this?

Could we tie what tests run against the way a PR has been tagged for instance? I find it's not overly needed to run all of the XML tests for instance against a PR which just (for arguments sake) changes some debug logging lines of code in a plugin.

Thanks,

Rob

**Comment by milliams** [comment]
I agree that we need to think about this. One big thing I've just tested locally to speed up `TestQueuedSubmit` is to change how the monitoring is done. Currently, if you look at the [test timings](https://ganga-ci.cern.ch/jenkins/job/PR%20builder/935/testReport/python.Ganga.test.GPI.Parallel.TestQueuedSubmit/TestQueuedSubmit/) you see that the vast majority of the time is spent waiting for the submitted tests to finish.

Currently in tests like this, the job monitoring is done via `GangaTest.Framework.utils.sleep_until_completed` which simply prompts the monitoring system to pick up the job and then do it's thing as normal. For tests like this where the monitoring is not really the thing that is being tested, I think that a simpler approach would be useful. I've written a small replacement for `sleep_until_completed` which looks something like the following (simplified):
```python
def run_until_state(j, timeout=60, sleep_period=1):
    end_time = datetime.utcnow() + timedelta(seconds=timeout)

    while j.status != 'completed' and datetime.utcnow() < end_time:
        type(j.backend).master_updateMonitoringInformation([j])
        time.sleep(sleep_period)

    return j.status == 'completed'
```
This runs all the monitoring for the job in the main thread and runs each of them serially.

Most importantly, switching to use this implementation takes `TestQueuedSubmit` from taking 6 minutes down to taking only 1 minute.

I can put together a PR with a few similar speed improvements for the tests?

**Comment by rob-c** [comment]
Then could we profile why it takes so long with the current test system? It's a bit big but that difference concerns me a little

**Comment by milliams** [comment]
It might be worth looking in to. I guess that it's mostly simply because there's so much more code in the monitoring system, doing things like intra-thread notification and blocking.

Making similar changes across the rest of the test suite (except for places that are testing the monitoring system), I have reduced the total time on my system of running the integration tests from about 21 minutes down to about 12 minutes (so far).

**Comment by drmarkwslater** [comment]
This sounds like a good thing to add. On a more general point, I plan to add a `@slow` decorator so any tests that are labelled as such will only be run on the Nightlies.

**Comment by rob-c** [comment]
OK, I'm curious. Say for instance I make a commit which breaks all of the slow tests which do something very time consuming but it passes all of the integration tests.

Is develop at this point broken?

If so, is it worth having a secondary branch of develop-stable (or whatever name) to pull from develop each night only when things are known to work?
I can think of a few reasons why this may be useful but I'm curious to hear people's opinions on this?
(it's easy for me to suggest this I'm likely not going to be the one setting it up)

**Comment by egede** [comment]
The idea is good but I am worried about complexity. At some point we will 
have to hand this onto other people and each time we increase complexity it 
gets harder.

Ulrik

**Comment by milliams** [comment]
It's possible (and common in industry) to have quite complex staging branches which feed into each other based on certain rules. For us I feel that git-flow is already complex enough.

To answer your question though @rob-c, indeed if we have certain tests demoted to be `external` or `slow` then they will not be automatically run on PRs and so it will be possible to have `develop` be 'broken'. There are two ways to reduce this problem:

1. Make sure that if you are working on a PR that you make sure that you run all the tests locally (at least the `slow` ones and hopefully the relevant `external` ones too). We have to trust each other on this.
2. Reduce the number of these tests. For slow tests there's not much we can do but for example for an external test which submits a DIRAC job, we should have an `external` one which actually communicates with the server but also an equivalent one which does exactly the same but with the DIRAC calls mocked away (or more sensibly, *one* test with a flag to select real DIRAC or mock).

**Comment by drmarkwslater** [comment]
@rob-c I think it's OK to have slow tests that we know are rock solid and should always pass pushed to Nightlies as though you're correct, we could break develop, it wouldn't be **completely** broken as the fast tests will always catch show stopping bugs straight away. Then, we we see the next day that the Nightlies are failing, we have to fix that ASAP and we don't release before they pass.

This brings up a subtle point though that we'd need to have the Release tests basically be the Nightlies (do we have release tests at the moment?) so we're sure we've run **every** test before a release.

**Comment by rob-c** [comment]
@milliams The tests which I'm likely to break only break on the Jenkins machines in practice and I'm keen to avoid ending up with the same problems as the old test system that the release manager _just knows_ to ignore test_xyz because it's been broken for a long time in testing.

I'm not a fan of trusting running the tests externally and if the tests are slow the onus is _still_ on the developer to run the tests on their machines which in my mind is hassle Jenkins was setup to reduce.
If this is the best solution then OK, I'm happy with this but I just think there may be a better approach.

@drmarkwslater  Given the level of discussion I'm not keen to implement _slow_ tests.

Having integration tests (albeit slow) and external nightly ones which may fail but require some thought seems enough of a distinction. Having slow tests in this case just adds more confusion and more work I think.

I would just like the ability to merge a PR which multiple people agree is not worth waiting on the tests to finish i.e. string formatting, documentation and alike.
i.e. stop enforcing the protected status of ```develop``` through software and have it as more of an agreement amongst developers.
Does this sound reasonable?

**Comment by drmarkwslater** [comment]
Actually, I'm **much** happier with `develop` being protected through software and I don't think this should change. You may have to wait 30mins for a minor doc string change to be OK'd but that's not the end of the world.

**Comment by rob-c** [comment]
Yes but it starts to get irritating when you have larger branches relying on smaller (harmless) changes which depend on another PR.
Other than speeding up the tests I'm not sure there's anything more we can do other than increase the complexity which does come at a too high price imo,
Shall I close this in favor of a new issue/PR for the tests optimisations?

**Comment by egede** [comment]
@rob-c Yes.

**Comment by rob-c** [comment]
Closing this and waiting for a PR on the tests being optimised...
